dll/win32/shell32/shelldesktop/CChangeNotify.cpp

```diff
@@ -0,0 +1,581 @@
+        ibOffset2 = DWORD_ALIGNMENT(ibOffset1 + cbPidl1);
+    }
+
+    cbSize = ibOffset2 + cbPidl2;
+    hTicket = SHAllocShared(NULL, cbSize, dwOwnerPID);
+    if (hTicket == NULL)
+    {
+        ERR("Out of memory\n");
+        return NULL;
+    }
+
+    pTicket = (LPDELITICKET)SHLockSharedEx(hTicket, dwOwnerPID, TRUE);
+    if (pTicket == NULL)
+    {
+        ERR("SHLockSharedEx failed\n");
+        SHFreeShared(hTicket, dwOwnerPID);
+        return NULL;
+    }
+    pTicket->dwMagic  = DELITICKET_MAGIC;
+    pTicket->wEventId = wEventId;
+    pTicket->uFlags = uFlags;
+    pTicket->ibOffset1 = ibOffset1;
+    pTicket->ibOffset2 = ibOffset2;
+
+    if (pidl1)
+        memcpy((LPBYTE)pTicket + ibOffset1, pidl1, cbPidl1);
+    if (pidl2)
+        memcpy((LPBYTE)pTicket + ibOffset2, pidl2, cbPidl2);
+
+    SHUnlockShared(pTicket);
+    return hTicket;
+}
+
+EXTERN_C LPHANDBAG
+DoGetHandBagFromTicket(HANDLE hTicket, DWORD dwOwnerPID)
+{
+    LPDELITICKET pTicket = (LPDELITICKET)SHLockSharedEx(hTicket, dwOwnerPID, FALSE);
+    if (!pTicket || pTicket->dwMagic != DELITICKET_MAGIC)
+    {
+        ERR("pTicket is invalid\n");
+        return NULL;
+    }
+
+    LPHANDBAG pHandBag = (LPHANDBAG)LocalAlloc(LMEM_FIXED, sizeof(HANDBAG));
+    if (pHandBag == NULL)
+    {
+        ERR("Out of memory\n");
+        SHUnlockShared(pTicket);
+        return NULL;
+    }
+    pHandBag->dwMagic = HANDBAG_MAGIC;
+    pHandBag->pTicket = pTicket;
+
+    pHandBag->pidl1 = pHandBag->pidl2 = NULL;
+    if (pTicket->ibOffset1)
+        pHandBag->pidl1 = (LPITEMIDLIST)((LPBYTE)pTicket + pTicket->ibOffset1);
+    if (pTicket->ibOffset2)
+        pHandBag->pidl2 = (LPITEMIDLIST)((LPBYTE)pTicket + pTicket->ibOffset2);
+
+    return pHandBag;
+}
+
+EXTERN_C void
+DoTransportChange(LONG wEventId, UINT uFlags, LPITEMIDLIST pidl1, LPITEMIDLIST pidl2,
+                  DWORD dwTick)
+{
+    HWND hwndNotif = DoGetNewDeliveryWorker();
+    if (!hwndNotif)
+        return;
+
+    DWORD pid;
+    GetWindowThreadProcessId(hwndNotif, &pid);
+
+    HANDLE hTicket = DoCreateDeliTicket(wEventId, uFlags, pidl1, pidl2, pid, dwTick);
+    if (hTicket)
+    {
+        TRACE("hTicket: %p, 0x%lx\n", hTicket, pid);
+        if ((uFlags & (SHCNF_FLUSH | SHCNF_FLUSHNOWAIT)) == SHCNF_FLUSH)
+            SendMessageW(hwndNotif, WM_NOTIF_DELIVERY, (WPARAM)hTicket, pid);
+        else
+            SendNotifyMessageW(hwndNotif, WM_NOTIF_DELIVERY, (WPARAM)hTicket, pid);
+    }
+}
+
+/*static*/ LRESULT CALLBACK
+CWorker::WindowProc(HWND hwnd, UINT uMsg, WPARAM wParam, LPARAM lParam)
+{
+    CWorker *pThis = (CWorker *)GetWindowLongPtrW(hwnd, 0);
+    if (pThis)
+    {
+        LRESULT lResult = 0;
+        pThis->ProcessWindowMessage(hwnd, uMsg, wParam, lParam, lResult, 0);
+        return lResult;
+    }
+    return 0;
+}
+
+BOOL CWorker::CreateWorker(HWND hwndParent, DWORD dwExStyle, DWORD dwStyle)
+{
+    m_hWnd = SHCreateWorkerWindowW(WindowProc, hwndParent, dwExStyle, dwStyle,
```

Comment:
assert that `m_hWnd` is null before overwriting it?

Reply:
Added assertion.